### PR TITLE
usb related commenting

### PR DESCRIPTION
**Pull request by frickler24**
This PR contains

- 99% commenting in Doxygen style (especially for the linux USB handling, even this topic is not complete yet)
- two Doxygen files (one for the whole documentation, one for the usb parts only)
- 2 static declarations of functions which are really static
- some changes in white spacing to enforce readability.

There is no functional change in the PR, so it would be helpful for me and maybe @fleischie (he wanted to start adding the macOS specific usb comments) if we had this in the testing stage.
@light2yellow or @tatokis please review.

### CODE REVIEW

**Comment by fleischie** [comment]
I actually did start, but not very much and then I got side-tracked a lot lately. If I find the time, I can continue with the macOS docs...

**Review by ghost — COMMENTED** [review]
Very good job, thanks.

Can you add the instructions on how to build the documentation? Maybe as well adding a target to qmake? Also, maybe build it out-of-source under doxygen/ and not inside the source directories?

**Comment by frickler24** [comment]
It is still WiP, trying the combination with travis and github-pages (gh-pages branch).
Interesting feature, a bit difficult to handle.

**Comment by frickler24** [comment]
Even if the PR was closed (due to the deletion of the branch, see below), 
I have taken a major step forward, and I ask you to discuss the result and, above all, the form of the creation.

In response to the feedback from @light2yellow  and because of the discussion in #126 and #84, I based the creation of the documentation on travis-ci.

You can see the result here in my [ckb-next github-pages](https://frickler24.github.io/ckb-next/index.html). The procedure has to be explained a bit:

First, the account is linked to github-pages and travis-ci is enabled to address the ckb-next repo. 
This causes a github.io page to be created in addition to the github.com page.

The gh-pages are initialized via another repo (currently called gh-pages-ckb-next in my account, it may be included in mattangers account), which contains some HTML/CSS files and two scripts for patching the links in the HTML files. The shell script for initialization is in ckb-next/documentation also.

With each PR and each push on the repo a travis-ci script is called with the name of the PR or branch as environment parameter. The behavior of the script  is configured in a[ .travis.yml file](https://travis-ci.org/frickler24/ckb-next/jobs/226606893/config).

In our example, I compile in a docker environment via quickinstall first in a [Debian / Ubuntu system](https://travis-ci.org/frickler24/ckb-next/jobs/226606893) and second in a [macOS configuration](https://travis-ci.org/frickler24/ckb-next/jobs/226606896). If the last compilation has finished, another shell script is called, in which the document creation is controlled. This shell script and some more files are located in the new folder "documentation" also.

The script loads the gh-pages branch into a subfolder, deletes any existing documentation for the current branch and runs Doxygen. Doxygen is currently running 3 times: 
- for the daemon only
- for the GUI client only and
- for the entire source code (daemon, GUI, scripts) 

to provide the appropriate documents for the particular application. If you have a look at the usage of globvars in the lightning scripts, you will understand the reason :-)
But this can all be matter of change.
BTW: The LaTex compilation to pdf has to fight against several UTF-8 special chars in the GUI (arrows, the appel-keysymbol etc.), so error messages from this make call are completely ignored. Maybe some of you have knowledge in LaTex to fix it.

An HTML version for online browsing, a PDF variant and a man(1) variant is generated.
Two scripts, which are located in the branch gh-pages, patch the links in the prepared HTML pages to direct to the newly generated documents.
Then the whole thing is committed and pushed. Github-pages automatically copies (or links?) the new content onto its pages.

The old branch for this PR had a "+" in the name, so I have deleted it for compatibility (URL names) and created a new one with the "+" replaced by a "-". 
Since I had to commit and push every time I tested the various scripts, the git history has become very long. If the result and the procedure via travis-ci gets your agreement, I make a patch to shorten the history (or is there a GIT command for this?).

So my first question is: Is the idea to use travis-ci at least for generating the docs OK for us?
If yes, I will create a PR with the scripts and a short description how to initialize the whole thing in the README.md

**Comment by fleischie** [comment]
Wow. First of all, thank you so much for your hard work. 😄

My thoughts:
- Maybe `lualatex`, `xelatex`, or any other newer implementation, as well as a different setup in the document might be able to compile the docs w/o warnings. I might take a look at this today.
- Using Travis might be useful not only for re-generating the docs but later as well to release a binary package and other CI/CD tasks as well, so I appreciate the pioneering in this section.
- I think you can squash-commit a branch, when merging, so all these commits will be accumulated in one single commit. (Edit: The other option is to rebase, but I think we squash commit.) -> so don't worry about the commit history, that's what's git for after all.

**Comment by fleischie** [comment]
I have checked out the PR branch and my more informed thoughts:
- First of all: I acknowledge, that this is a WIP.
- The setup is a little messy, I had trouble understanding the relations between each file and dir (i.e. `.travis.yml`, `documentation/`, etc.)
- I had troubles with the doxygen-generation as well. It depends heavily on global variables, which I couldn't reproduce. Maybe a separation between the doxygen-generation and deployment might help here.
- Using `lulatex` instead of `pdflatex` in the generated `Makefile` does indeed work without problems, but the appropriate characters are not displayed at all in the resulting PDF. (e.g. it says: "Command ()", instead of "Command (⌘)").

So, to summarize: It needs some work to be truly integrated well into the workflow but it's a solid foundation to work on, both in terms of documentation as well as CI/CD. I am looking forward to this being integrated. :+1: :slightly_smiling_face:

**Comment by frickler24** [comment]
@fleischie Thanks for the review and hints!
First, I will try to create a picture to explain the relationships and interdependencies of the various configfiles and directories. This is really confusing and maybe there is a simpler approach to this.
The image will also help to understand the role of the global environment variables that come from travis-ci itself and control the doxygen process.
And yes - I also believe that with a little more intensive work with the tool we can get mechanisms to create fully generated installation packages.
BTW: Automatic testing is hardly possible, because without corresponding hardware the daemon will find nothing and thus the GUI will not display anything. I have tried to have the daemon installed in the Dockercontainer, but there is no systemd in the installation. The call with sudo works in principle, but then the devices are missing.
Before I put more time and energy into the topic, I wanted to collect your feedback. Especially @mattanger has to agree, that in his account travis-ci is used (which can only be adjusted for the account and then the tracked repos can be de-/selected).

**Comment by fleischie** [comment]
You're welcome. 😊

OK, no auto-tests. But Packaging + Release on succesful compilation might be sufficient for the most general cases. It's not crucial at this point but nice to have perspectively. 😄

**Comment by frickler24** [comment]
@fleischie here is the [first graphic](https://www.draw.io/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&title=Doxygen%20via%20Travis-ci.xml#Uhttps%3A%2F%2Fraw.githubusercontent.com%2Ffrickler24%2Fdiverses%2Fmaster%2FDoxygen%2520via%2520Travis-ci.xml) for the generation  process.
While I've done dis documentation, I have seen, that a lot of files have come to branches which should not be there. I'll clean this the next days, currently we have a big release "in real life" at weekend up to monday.
Maybe tomorrow I'll find some time to document the initialization process.

**Comment by fleischie** [comment]
OK, thanks for the info. I will try to keep myself updated on this. 😄

**Comment by frickler24** [comment]
And here is the [documentation of the installation script](https://www.draw.io/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&title=Initialize%20doxygen%20via%20Travis-ci.xml#Uhttps%3A%2F%2Fraw.githubusercontent.com%2Ffrickler24%2Fdiverses%2Fmaster%2FInitialize%2520doxygen%2520via%2520Travis-ci.xml).
